### PR TITLE
Use namespaced Twig

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | needs https://github.com/twigphp/Twig/pull/2490 and https://github.com/twigphp/Twig/pull/2491
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Bridge/Twig/Command/LintCommand.php [hunk 1]
```diff
@@ -17,6 +17,11 @@
 use Symfony\Component\Console\Input\InputOption;
 use Symfony\Component\Console\Output\OutputInterface;
 use Symfony\Component\Finder\Finder;
+use Twig\Environment;
+use Twig\Error\Error;
+use Twig\Loader\ArrayLoader;
+use Twig\Source;
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Command/LintCommand.php]: @keradus just wondering: this `use`is not used in the class - is it a php-cs-fixer bug that fabbot is still green?

File: src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php [hunk 2]
```diff
@@ -15,6 +15,10 @@
 use Symfony\Component\HttpKernel\DataCollector\LateDataCollectorInterface;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
+use Twig\Markup;
+use Twig\Profiler\Dumper\HtmlDumper;
+use Twig\Profiler\Profile;
+use Twig\Source;
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php]: @keradus just wondering: this `use`is not used in the class - is it a php-cs-fixer bug that fabbot is still green?

File: src/Symfony/Bridge/Twig/Command/DebugCommand.php [hunk 3]
```diff
@@ -37,15 +38,15 @@ public function __construct($name = 'debug:twig')
     /**
      * Sets the twig environment.
      *
-     * @param \Twig_Environment $twig
+     * @param Environment $twig
      */
-    public function setTwigEnvironment(\Twig_Environment $twig)
+    public function setTwigEnvironment(Environment $twig)
```
Comment (fabpot) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: Reading this, I would rename `Twig\Environment` to `Twig\Twig`

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Twig/Command/DebugCommand.php (hunk 3 above)
Comment (javiereguiluz) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: Is it possible to name it just `Twig` to avoid the `Twig\Twig` repetition?

File: src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php (hunk 2 above)
Comment (jvasseur) [inline src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php]: The `NoUnusedImportsFixer` is quite permissive (not removing an unsed import in better than removing an used one) particularly it check if the class in not referenced in a comment. In this case it think it's the "source" word in the license header comment that make the fixer think that the class is still used.

File: src/Symfony/Bridge/Twig/Command/DebugCommand.php (hunk 3 above)
Comment (Simperfit) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: Yep but you will have to import Twig\Twig anyway

Comment (fabpot) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: We cannot use `Twig` only as all classes need to be under a namespace. That's PSR-4.

Comment (GromNaN) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: @javiereguiluz This would not be PSR-4 compliant. In §2.2.1, "The fully qualified class name MUST have a top-level namespace name, also known as a “vendor namespace”.
http://www.php-fig.org/psr/psr-4/#specification

File: src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php (hunk 2 above)
Comment (keradus) [inline src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php]: that's exactly the case here.
Imported class name could be used in comment, even in different casing, which later is utilized by Doctrine or some annotation-parsers. that behaviour was added after fixer broke some code.
I would not focus here is it header comment or other comment actually.

One could bring configuration to fixer to ignore comments or non-docblock comments / be strict about casing !

Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php]: Is there a way to make it more strict so that I can cleanup locally?

File: src/Symfony/Bridge/Twig/Command/DebugCommand.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: renamed to Twig\Twig

File: src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php (hunk 2 above)
Comment (julienfalque) [inline src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php]: @nicolas-grekas Not currently, see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/2814.

Comment (keradus) [inline src/Symfony/Bridge/Twig/DataCollector/TwigDataCollector.php]: go ahead @nicolas-grekas and contribute ;)

**Comment by nicolas-grekas** [comment]
Just fixed a few potential BC breaks related to the use of aliases type hints.
Should be ready, waiting for https://github.com/twigphp/Twig/pull/2490

File: src/Symfony/Bridge/Twig/Command/DebugCommand.php [hunk 4]
```diff
@@ -165,7 +161,7 @@ private function getMetadata($type, $entity)
                     return false;
                 }
 
-                return !$param->getClass() || $param->getClass()->getName() !== 'Twig_Environment';
+                return !$param->getClass() || !in_array($param->getClass()->getName(), array('Twig_Environment', 'Twig\Twig'));
```
Comment (GromNaN) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: Using `class_alias`, the `ReflectionClass` refers to the original class. Checking against `Twig\Twig` is not necessary.
https://3v4l.org/QH2KQ

Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: True, until Twig moves to namespaces-first, which is why this code is here.

Comment (stof) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: Actually, this logic could be improved to avoid relying on typehints or param names (btw, the check on param name is wrong, as the name is not relevant). I will send a PR.

Comment (stof) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: here you go: https://github.com/symfony/symfony/pull/22994

File: src/Symfony/Bridge/Twig/Extension/RoutingExtension.php [hunk 5]
```diff
@@ -82,9 +87,11 @@ public function getUrl($name, $parameters = array(), $schemeRelative = false)
      * - path('route', {'param1': 'value1', 'param2': 'value2'})
      * If param1 and param2 reference placeholder in the route, it would still be safe. But we don't know.
      *
-     * @param \Twig_Node $argsNode The arguments of the path/url function
+     * @param Node $argsNode The arguments of the path/url function
      *
      * @return array An array with the contexts the URL is safe
+     *
+     * To be made @final in 3.4, and the type-hint be changed to "\Twig\Node\Node" in 4.0.
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Extension/RoutingExtension.php]: Wouldn't this trigger the warning already ? Or does DebugClassLoader check for beginning of lines ?

File: src/Symfony/Bridge/Twig/Extension/RoutingExtension.php [hunk 6]
```diff
@@ -82,9 +87,11 @@ public function getUrl($name, $parameters = array(), $schemeRelative = false)
...
      */
     public function isUrlGenerationSafe(\Twig_Node $argsNode)
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Extension/RoutingExtension.php]: Can't we already change the typehint here (with the same `class_exists` call) ?

File: src/Symfony/Bridge/Twig/Command/DebugCommand.php (hunk 4 above)
Comment (stof) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: @nicolas-grekas please revert this change to avoid useless conflicts with my PR removing this code entirely 😄

File: src/Symfony/Bridge/Twig/Extension/RoutingExtension.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Extension/RoutingExtension.php]: it's ok: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Debug/DebugClassLoader.php#L171

File: src/Symfony/Bridge/Twig/Extension/RoutingExtension.php (hunk 6 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Extension/RoutingExtension.php]: We can, but that means loading Node, which in turn loads Compiler. I preferred not, thus the current code.

File: src/Symfony/Bridge/Twig/Command/DebugCommand.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Twig/Command/DebugCommand.php]: reverted, thanks

**Comment by nicolas-grekas** [comment]
now green

**Review by fabpot — APPROVED** [review]
You just need to revert the `Twig\Environment` => `Twig\Twig` change. Sorry about that :)

**Comment by nicolas-grekas** [comment]
renaming done, now green
